### PR TITLE
Refactor/new implementation

### LINKED ISSUE #38: Better implementation
@riderodd I am facing a problem where the `onFinalResult` event is never triggered, and I believe this change is the reason since everything is cleaned up on the first result. Does it make sense?

_Originally posted by @joazvsoares in https://github.com/riderodd/react-native-vosk/pull/32#discussion_r1411547254_

### PR DESCRIPTION
**Pull request by kingdcreations**
close #38

### CODE REVIEW

**Review by riderodd — CHANGES_REQUESTED** [review]
Nice work @kingdcreations , can you update the Readme with a warning about this breaking change and I take care of the iOS part ?

**Comment by riderodd** [comment]
@kingdcreations : how do you specify an array of available strings for the recognition now? (the old grammar parameter)

**Comment by kingdcreations** [comment]
> @kingdcreations : how do you specify an array of available strings for the recognition now? (the old grammar parameter)

No changes !

> Nice work @kingdcreations , can you update the Readme with a warning about this breaking change and I take care of the iOS part ?

I updated it with new functionalities, but I'd still like to add those:

* Timeout handling, I want to add an argument to `start()` but as there is no function overloading in js, should i set `start()` args to an options object like: 
```js
start({
 grammar: ["left", "right", "[unk]"],
 timeout: 10000
})
 ```
 
 * String promises instead of { data: String } actual one
 
 What do you think ?

**Comment by riderodd** [comment]
I'm loving it!

Can you just add a test on the example app with grammar and one without ?

**Comment by kingdcreations** [comment]
All done ! However react native 0.73 is here, should i update ?

**Comment by riderodd** [comment]
> All done ! However react native 0.73 is here, should i update ?

No let's keep that for later

File: README.md [hunk 1]
```diff
@@ -15,18 +15,16 @@ Avoid using too heavy models, because the computation time required to load them
...
 
-Then put the model folder inside the assets folder created. In your file tree it should be located in android\app\src\main\assets. So, if you downloaded the french model named model-fr-fr, you should access the model by going to android\app\src\main\assets\model-fr-fr. In Android studio, your project structure should be like that:
 ![Android Studio final project structure](https://raw.githubusercontent.com/riderodd/react-native-vosk/main/docs/android_studio_project_structure.png)
 
 You can import as many models as you want.
 
 ### iOS
-In XCode, right-click on your project folder, and click on "Add files to [your project name]".
-
-![XCode add files to project](https://raw.githubusercontent.com/riderodd/react-native-vosk/main/docs/xcode_add_files_to_folder.png)
```
Comment (riderodd) [inline README.md:29]: Why removing the picture ?

File: README.md [hunk 2]
```diff
@@ -15,18 +15,16 @@ Avoid using too heavy models, because the computation time required to load them
 Then, unzip the model in your app folder. If you just need to use the iOS version, put the model folder wherever you want, and import it as described below. If you need both iOS and Android to work, you can avoid to copy the model twice for both projects by importing the model from the Android assets folder in XCode. Just do as follow:
 
 ### Android
-In Android Studio, open the project manager, right-click on your project folder and New > Folder > Assets folder.
-![Android Studio assets folder creation](https://raw.githubusercontent.com/riderodd/react-native-vosk/main/docs/android_studio_assets_folder_creation.png)
```
Comment (riderodd) [inline README.md:19]: Shouldn't we keep the picture ?

File: README.md [hunk 3]
```diff
@@ -43,67 +41,130 @@ import Vosk from 'react-native-vosk';
...
+| `onTimeout` | `void` | Called after timeout expired |
+
+### Examples
+
+#### Default
+
+```js
+const start = () => vosk.start();
+
+useEffect(() => {
```
Comment (riderodd) [inline README.md]: useEffect can be misinterpreted here, I think we should show a more agnostic example, don't you ?

File: README.md [hunk 4]
```diff
@@ -43,67 +41,130 @@ import Vosk from 'react-native-vosk';
...
+  });
+
+  return () => resultEvent.remove();
+}, [vosk]);
+```
+
+#### Using grammar
+
+```js
+const start = () =>
```
Comment (riderodd) [inline README.md]: vosk.start() returns a promise, I think it's not clear enough here. What about:
```js
vosk.start({ grammar: [...] }).then(() => {
const resultEvent = vosk.onResult((res) => {
    if (res === 'left') {
      console.log('Go left');
    } else if (res === 'right') {
      console.log('Go right');
    } else {
      console.log("Instruction couldn't be recognized");
    }
  });

// when done, remember to call resultEvent.remove();
})
```

**Review by riderodd — COMMENTED** [review]
Amazing work thanks !
I'm a bit confused about all the events, but that's the way Vosk works so... If you have any suggestion to make it clearer feel free to make a suggestion !
As soon as everything is resolved I'll work on the iOS side.

File: README.md (hunk 1 above)
Comment (kingdcreations) [reply README.md:29]: I felt like the installation guide was too long but I can put it back !

Comment (kingdcreations) [reply README.md:29]: Also I'd like to have a js based loader, would be easier to install and maintain, like:

```js
import modelEN from "models/model-en"

vosk.load(modelEN)
```

But I don't know if that even possible, wdy think ?

File: README.md (hunk 2 above)
Comment (kingdcreations) [reply README.md:19]: You tell me ;)

File: README.md (hunk 1 above)
Comment (kingdcreations) [reply README.md:29]: I added back images for now !

Comment (riderodd) [reply README.md:29]: > Also I'd like to have a js based loader, would be easier to install and maintain, like:
> 
> ```js
> import modelEN from "models/model-en"
> 
> vosk.load(modelEN)
> ```
> 
> But I don't know if that even possible, wdy think ?

Unfortunately no, it's not. At least when I first researched how to include a file in the final bundle, maybe we can find a way now, but some months ago I didn't figured it out.

**Review by riderodd — COMMENTED** [review]
Nice thanks !

**Comment by riderodd** [comment]
@kingdcreations : would you review my last commits ?

File: ios/Vosk.swift [hunk 5]
```diff
@@ -69,25 +87,57 @@ class Vosk: RCTEventEmitter {
...
+            resolve(true);
+        } catch {
+            reject(nil, nil, nil);
+        }
+    }
+
+
+    /// Set grammar to use
+    @objc(setGrammar:withResolver:withRejecter:)
+    func setGrammar(words: [String]?, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
```
Comment (kingdcreations) [inline ios/Vosk.swift]: No cleaning/ restart needed ? Even after multiple calls/ recognition ? No leak ?

File: ios/Vosk.swift [hunk 6]
```diff
@@ -117,16 +170,23 @@ class Vosk: RCTEventEmitter {
                 try? self.audioEngine.start()
             }
             
-            // and manage timeout
-            timeoutTimer = Timer.scheduledTimer(withTimeInterval: 15, repeats: false) {_ in
-                self.sendEvent(withName: "onTimeout", body: ["data": ""])
-                self.stopInternal(withoutEvents: true)
+            // if grammar is set in options, override the current grammar
```
Comment (kingdcreations) [inline ios/Vosk.swift]: Wrong description ?

File: ios/Vosk.swift (hunk 5 above)
Comment (riderodd) [reply ios/Vosk.swift]: Grammar is stored into an internal class variable, for sure it won't work unless you call stop() and start() again. How did you do on Android ? I may have implemented a wrong pattern ^^

File: ios/Vosk.swift (hunk 6 above)
Comment (riderodd) [reply ios/Vosk.swift]: Fixed in last commit, thanks !

File: ios/Vosk.swift (hunk 5 above)
Comment (kingdcreations) [reply ios/Vosk.swift]: I clean the recognizer, call the native setGrammar method, and restart, wdy think ?

```kotlin
@ReactMethod
  fun setGrammar(grammar: ReadableArray? = null, promise: Promise) {
    if (recognizer == null || speechService == null) {
      promise.reject(IOException("Recognizer is not started yet"))
    } else {
      try {
        speechService!!.stop()
        speechService!!.shutdown()

        if (grammar != null)
          recognizer!!.setGrammar(makeGrammar(grammar))
        else
          recognizer!!.setGrammar("[]")

        speechService = SpeechService(recognizer, sampleRate)
        if (speechService!!.startListening(this))
          return promise.resolve("Recognizer successfully started")

      } catch (e: IOException) {
        promise.reject(e)
      }
    }
  }
```

Comment (riderodd) [reply ios/Vosk.swift]: @kingdcreations : ok I did the same implementation on my side, but it made me think about the timeout argument that id not preserved, isn't it ? I mean, I call start with timeout: 5, then setGrammar(), and my timeout disappear no ?

Comment (kingdcreations) [reply ios/Vosk.swift]: You are right, what should i do ?

Comment (riderodd) [reply ios/Vosk.swift]: In a first time, removing setGrammar seem fine no ? Devs can just call stop() and start() again with grammar in parameters, and they keep control of the recognition state on their UI. Isn't it better?

Comment (kingdcreations) [reply ios/Vosk.swift]: Yup, lets do that

**Review by riderodd — APPROVED** [review]

**Pull request merged**